ext/drreg/drreg.c

```diff
@@ -385,14 +385,20 @@ drreg_event_bb_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_tr
             /* DRi#1849: COND_SRCS here includes addressing regs in dsts */
             if (instr_reads_from_reg(inst, reg, DR_QUERY_INCLUDE_COND_SRCS))
                 value = REG_LIVE;
-            /* make sure we don't consider writes to sub-regs */
-            else if (instr_writes_to_exact_reg(inst, reg, DR_QUERY_INCLUDE_COND_SRCS)
-                     /* A write to a 32-bit reg zeroes the top 32 bits for x86_64 and
-                      * aarch64.
-                      */
-                     IF_X64(||
-                            instr_writes_to_exact_reg(inst, reg_64_to_32(reg),
-                                                      DR_QUERY_INCLUDE_COND_SRCS)))
+            /* Make sure we don't consider writes to sub-regs. i#6417: in case
+             * of a syscall, restore the value of the output register since it
```

Comment:
I'm actually a little confused as to whether a trace will be built through a non-ignorable syscall (drmemtrace marks all syscalls as non-ignorable via the filter event) -- so I'm not certain what happens without `-disable_traces` in drmemtrace.  I would think it can't build the trace as it needs to invoke the client callbacks; but at a glance I don't see FRAG_MUST_END_TRACE being added.  Probably just missed it.

Reply:
I've updated the code to treat syscall as a barrier to restore register values. Please take another look,